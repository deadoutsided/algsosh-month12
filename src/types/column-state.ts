import { ElementStates } from "./element-states";

export type TColumnState = {
  num: number;
  state: ElementStates;
  id: number;
};
