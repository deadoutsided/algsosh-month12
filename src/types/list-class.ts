export class ListItem<T> {
  value: T;
  next: ListItem<T> | null;
  constructor(value: T, next?: ListItem<T> | null) {
    this.value = value;
    this.next = next !== undefined ? next : null;
  }
}

interface ILinkedList<T> {
  append: (element: T, index: number) => void;
  delete: (index: number) => void;
  getSize: () => number;
  getHead: () => ListItem<T> | null;
}

export class LinkedList<T> implements ILinkedList<T> {
  private head: ListItem<T> | null;
  private tail: ListItem<T> | null;
  private size: number;
  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  append(element: T, index: number) {
    if (index > this.size || index < 0) {
      throw new Error("Incorrect index");
    }
    let temp = new ListItem<T>(element);
    if (index === 0) {
      console.log(this.head);
      temp.next = this.head;
      console.log(this.head);
      this.head = temp;
      console.log(this.head);
    } else {
      let curr = this.head;
      let currIndex = 0;
      let changed: ListItem<T> | null = null;
      while (curr && currIndex < index - 1) {
        curr = curr.next;
        currIndex++;
      }
      if (curr && currIndex === index - 1) {
        changed = curr.next;
      }
      temp.next = changed;
      if (curr) curr.next = temp;
    }
    if (index === this.size) {
      this.tail = temp;
    }
    this.size++;
  }

  delete(index: number) {
    if (index > this.size || index < 0 || !this.head) {
      throw new Error("Incorrect index");
    }
    let temp = this.head;
    let tempInd = 0;
    while (temp.next && tempInd !== index - 1) {
      temp = temp.next;
      tempInd++;
    }
    if (temp && tempInd === index - 1) {
      let delTarg = temp.next;
      temp.next = delTarg && delTarg.next ? delTarg.next : null;
    }
    if (index === 0 && this.head !== null) {
      this.head = this.head.next ? this.head.next : null;
    }
    this.size--;
    if (temp.next === null) {
      this.tail = temp;
    }
  }

  getSize() {
    return this.size;
  }

  getHead() {
    return this.head;
  }
}
