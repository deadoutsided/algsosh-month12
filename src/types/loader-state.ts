export type TLoader = {
  constDelLoaderHead: boolean;
  constDelLoaderTail: boolean;
  indexDelLoader: boolean;
  indexAddLoader: boolean;
  constAddLoaderHead: boolean;
  constAddLoaderTail: boolean;
}