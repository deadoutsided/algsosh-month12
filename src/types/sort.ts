export enum SortType {
  Bubble = "bubble",
  Choose = "choose",
}
