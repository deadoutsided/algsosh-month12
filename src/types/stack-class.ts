interface IStack<T> {
  push: (item: T) => void;
  pop: () => void;
  clear: () => void;
  getSize: () => number;
  getContainer: () => T[];
  getTop: () => T | null;
}

export class Stack<T> implements IStack<T> {
  private container: T[];
  private top: T | null;
  constructor() {
    this.container = [];
    this.top = null;
  }

  push = (item: T) => {
    this.container.push(item);
  };

  pop = () => {
    this.container.pop();
  };

  clear = () => {
    this.container = [];
  };

  getTop = () => {
    const topElem =
      this.getSize() > 0 ? this.container[this.getSize() - 1] : null;
    return topElem;
  };

  getSize = () => this.container.length;

  getContainer = () => this.container;
}
