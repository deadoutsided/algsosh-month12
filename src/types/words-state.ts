import { ElementStates } from "./element-states";

export type TWord = {
  word: string;
  id: number | string;
  state?: ElementStates;
  head?: string | TWord;
  tail?: string | TWord;
};
