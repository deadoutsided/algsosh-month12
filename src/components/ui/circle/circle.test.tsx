import React from "react";
import renderer from 'react-test-renderer';

import { Circle } from "./circle";
import { ElementStates } from "../../../types/element-states";

describe('Circle component', () => {
  it('hollow circle', () => {
    const tree = renderer.create(<Circle />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with letter', () => {
    const tree = renderer.create(<Circle letter="a" />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with letter in head', () => {
    const tree = renderer.create(<Circle head="1"/>).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with react elem in head', () => {
    const tree = renderer.create(<Circle head={<Circle letter="a" isSmall={true} />} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with letter in tail', () => {
    const tree = renderer.create(<Circle tail="a" />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with react elem in tail', () => {
    const tree = renderer.create(<Circle tail={<Circle letter="a" isSmall={true} />} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with index', () => {
    const tree = renderer.create(<Circle index={15} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with isSmall = true', () => {
    const tree = renderer.create(<Circle isSmall={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with state default', () => {
    const tree = renderer.create(<Circle state={ElementStates.Default} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with state changing', () => {
    const tree = renderer.create(<Circle state={ElementStates.Changing} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('circle with state modified', () => {
    const tree = renderer.create(<Circle state={ElementStates.Modified} />).toJSON();
    expect(tree).toMatchSnapshot();
  })
})