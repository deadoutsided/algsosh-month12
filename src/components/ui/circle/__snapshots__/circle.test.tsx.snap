// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle component circle with index 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  >
    15
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail60 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with isSmall = true 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle  small default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with letter 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      a
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with letter in head 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
    data-testid="circleHead"
  >
    1
  </div>
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with letter in tail 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
    data-testid="circleTail"
  >
    a
  </div>
</div>
`;

exports[`Circle component circle with react elem in head 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  >
    <div
      className="content "
      data-testid="circle"
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
        data-testid="circleHead"
      />
      <div
        className="circle  small default"
        data-testid="circleInsides"
      >
        <p
          className="text text_type_circle text_color_input letter"
        >
          a
        </p>
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
        data-testid="circleIndex"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
        data-testid="circleTail"
      />
    </div>
  </div>
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with react elem in tail 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  >
    <div
      className="content "
      data-testid="circle"
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
        data-testid="circleHead"
      />
      <div
        className="circle  small default"
        data-testid="circleInsides"
      >
        <p
          className="text text_type_circle text_color_input letter"
        >
          a
        </p>
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
        data-testid="circleIndex"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
        data-testid="circleTail"
      />
    </div>
  </div>
</div>
`;

exports[`Circle component circle with state changing 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   changing"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with state default 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component circle with state modified 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   modified"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;

exports[`Circle component hollow circle 1`] = `
<div
  className="content "
  data-testid="circle"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
    data-testid="circleHead"
  />
  <div
    className="circle   default"
    data-testid="circleInsides"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
    data-testid="circleIndex"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="circleTail"
  />
</div>
`;
